openvdb/points/AttributeArray.h

```diff
@@ -836,6 +1019,92 @@ UnitVecCodec::encode(const math::Vec3<T>& val, StorageType& data)
 }
 
 
+////////////////////////////////////////
+
+// AttributeArray implementation
+
+#if OPENVDB_ABI_VERSION_NUMBER >= 6
+template <typename IterT>
+void AttributeArray::copyValuesUnsafe(const AttributeArray& sourceArray, const IterT& iter,
+    bool rangeChecking/* = false*/)
+{
+    // ensure both arrays have float-float or integer-integer value types
+    assert(sourceArray.valueTypeIsFloatingPoint() != this->valueTypeIsFloatingPoint());
+    // ensure both arrays have been loaded from disk (if delay-loaded)
+    assert(sourceArray.isDataLoaded() && this->isDataLoaded());
+    // ensure storage size * stride matches on both arrays
+    assert(this->storageTypeSize()*this->stride() == sourceArray.storageTypeSize()*sourceArray.stride());
+
+    const size_t bytes(sourceArray.storageTypeSize()*sourceArray.stride());
+    const char* const sourceBuffer = sourceArray.dataAsByteArray();
+    char* const targetBuffer = this->dataAsByteArray();
+    assert(sourceBuffer && targetBuffer);
+
+    if (rangeChecking && this->isUniform()) {
+        OPENVDB_THROW(IndexError, "Cannot copy array data as target array is uniform.");
+    }
+
+    const bool sourceIsUniform = sourceArray.isUniform();
+
+    const Index sourceDataSize = rangeChecking ? sourceArray.dataSize() : 0;
+    const Index targetDataSize = rangeChecking ? this->dataSize() : 0;
+
+    for (IterT it(iter); it; ++it) {
+        const Index sourceIndex = sourceIsUniform ? 0 : it.sourceIndex();
+        const Index targetIndex = it.targetIndex();
+
+        if (rangeChecking) {
+            if (sourceIndex >= sourceDataSize) {
+                OPENVDB_THROW(IndexError,
+                    "Cannot copy array data as source index exceeds size of source array.");
+            }
+            if (targetIndex >= targetDataSize) {
+                OPENVDB_THROW(IndexError,
+                    "Cannot copy array data as target index exceeds size of target array.");
+            }
```

Comment:
This feels a bit clunky, especially in an "unsafe" method. Is there no way we can promote this rangeChecking flag to copyValues() which would check the iterator extrema?

Reply:
Okay well two suggestions then:

1) Let the user deal with this in their iterator. Update the example iterator to store two size_t sizes for the source and target arrays and show range checking in the iterator. Completely remove everything bar the asserts.
2) Define functionality for out_of_range values which doesn't involve a throw. Maybe just ```if (sourceIndex >= sourceDataSize) continue;```

I think the first is the best solution

Reply:
I'm not sure I agree with your proposed solution as I don't like the idea that the user has to provide the minimum and maximum values. If the indices are not predictable and not in order that means either a sort or a second pass through the data to figure that out, just in order to do range checking. The range checking _is_ done automatically, but is done on every index. If you _really_ care about this performance the unsafe method is offered to eliminate that cost, provided you know what you're doing.

I do however agree that the range-checking parameter on the unsafe method is redundant, which bloats the public API slightly. As a result, I've introduced a new private member function which performs the copying and the range-checking parameter and both copyValues() and copyValuesUnsafe() call that method with the appropriate value.

Reply:
The user only needs to provide at most the two sizes of the arrays from dataSize(), not the minimum and maximum values. What I'm proposing is not performing any additional pass to do range checking. Its not necessarily about performance but also about what an "unsafe" method should represent. imo it shouldn't throw. I'm suggesting that a user can provide something like this (taken from your unit tests):

```
struct MyIterator
{
    using T = std::vector<std::pair<Index, Index>>;

    VectorWrapper(const T& _data, Index max) : data(_data), mMax(max) { }
    operator bool() const { return index < data.size(); }
    VectorWrapper& operator++() { index++; return *this; }
    Index sourceIndex() const { assert(*this); 
        if (data[index].first >= mMax) throw std::runtime_error("out_of_range"); // throw
        return data[index].first; 
    }
    Index targetIndex() const { assert(*this); 
        if (data[index].second >= mMax) return 0; // do some custom thing
        return data[index].second; 
    }

private:
    const T& data;
    const Index mMax;
    T::size_type index = 0;
};
```
simplifies the API, no additional pass, defers the control to the user as to how to handle out of range values in all cases, allows for optimal performance in the general case.

Reply:
The point is **not** to defer to the user to implement their own bounds-checking and hope they do it correctly, this _safe_ method should guarantee to **never** result in a seg-fault regardless of what array data or input you call it with. Consider std::vector::at() which does bounds-checking vs std::vector[] which doesn't for example. 

Since I've removed the range-checking parameter from the _unsafe_ method, this method does **not** throw.

We could provide just one implementation of this method with various different toggles for different permutations of checking ranges, strided sizes, value type checking, auto-expanding, throwing or asserting on errors, etc, etc. I think this overly complicates the API, hence the preferred balance between flexibility, simplicity, safety and performance in my opinion is a _safe_ method which provides all guarantees and checks and throws exceptions on errors and an _unsafe_ method which has none of these guarantees but offers bare-metal performance with asserts to help track down seg-faults.

Reply:
Right I didn't realise your change removed the issue that the unsafe method could throw, that's great! In which case we're left with the unnecessary branch complexity (the difference with your at vs vec analogy...). I am not suggesting providing a method for each permutation :\ yes this would definitely over-complicate the API.... I'm making the case that its better overall to defer this logic to the user (where they could define this). The 'hope they do it correctly' statement is a bit weird - you're designing an API to be used in a certain way, you're allowed to define that - there's a number of weird ways I could call copyValuesUnsafe with a custom object to cause problems (what if I don't increment, should we catch that? clearly not!). Obviously there's a fine line but I guess it's personal preference.

Reply:
True, you can't protect for everything. I suppose I see it as most important to protect the user against accidentally introducing undefined behavior when using the safe implementation than introducing an infinite loop.